cmd/github.go

```diff
@@ -0,0 +1,351 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"math/rand"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/juju/ansiterm"
+	"github.com/skratchdot/open-golang/open"
+	"github.com/urfave/cli"
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/github"
+
+	"github.com/manifoldco/manifold-cli/analytics"
+	"github.com/manifoldco/manifold-cli/api"
+	"github.com/manifoldco/manifold-cli/color"
+	"github.com/manifoldco/manifold-cli/config"
+	"github.com/manifoldco/manifold-cli/generated/identity/client/authentication"
+	"github.com/manifoldco/manifold-cli/generated/identity/client/user"
+	"github.com/manifoldco/manifold-cli/generated/identity/models"
+	"github.com/manifoldco/manifold-cli/prompts"
+)
+
+var (
+	sourceGitHub      = "github"
+	sourceGitHubToken = "github-token"
+	stateChars        = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ123456789"
+)
+
+// githubAuthorizationRequest is the request sent to GitHub for Personal Authorization Tokens
+type githubAuthorizationRequest struct {
+	Scopes      []string `json:"scopes"`
+	Notes       string   `json:"note"`
+	Fingerprint string   `json:"fingerprint"`
+}
+
+// githubAuthorizationResponse is the response that is given for Personal Authorization Token
+// requests
+type githubAuthorizationResponse struct {
+	ID    int
+	Token string
+
+	Errors []struct {
+		Code string
+	}
+}
+
+// githubErrors are the human readable error messages from GitHub during the Personal Authorization
+// Token flow
+var githubErrors = map[string]string{
+	"already_exists": "A Personal Authorization Token already exists for Manifold",
+}
+
+// oauthStoreFunc is responsible for storing the OAuth authentication with Manifold, either by
+// linking accounts or logging in
+type oauthStoreFunc func(ctx context.Context, cfg *config.Config, a *analytics.Analytics, req models.OAuthAuthenticationRequest) error
+
+// githubWithCallback identified with GitHub via. the normal browser-based OAuth flow
+func githubWithCallback(ctx context.Context, cfg *config.Config, a *analytics.Analytics, store oauthStoreFunc) error {
+	authConfig := &oauth2.Config{
+		ClientID:    config.GitHubClientID,
```

Comment:
We should document the new configuration and callback URL in the README or markdown somewhere.

Reply:
Like a "how to setup your oauth app"

Reply:
Added this: https://github.com/manifoldco/manifold-cli/blob/0d6d4f270844dadf4db43dc983ee6a63b69b8f6b/README.md#oauth-identity though one issue I see is the client ID we have on the backend marketplace has to match with the one the original exchange is done with, so external users can't really change their configuration without breaking their authentication